closed/src/java.base/share/classes/openj9/internal/criu/NativePRNG.java

```diff
@@ -239,11 +273,100 @@ public OutputStream run() {
                     try {
                         seedOut.write(seed);
                     } catch (IOException e) {
-                        // Ignored. On Mac OS X, /dev/urandom can be opened
-                        // for write, but actual write is not permitted.
+                        // ignored, On Mac OS X, /dev/urandom can be opened
+                        // for write, but actual write is not permitted
                     }
                 }
             }
         }
+
+        // ensure that there is at least one valid byte in the buffer
+        // if not, read new bytes
+        private void ensureBufferValid() throws IOException {
+            long time = System.currentTimeMillis();
+            int new_buffer_size = 0;
+
+            // check if buffer has bytes available that are not too old
+            if (buffered > 0) {
+                if ((time - lastRead) < MAX_BUFFER_TIME) {
+                    return;
+                } else {
+                    // byte is old, so subtract from counter to shrink buffer
+                    change_buffer--;
+                }
+            } else {
+                // no bytes available, so add to count to increase buffer
+                change_buffer++;
+            }
+
+            // if counter has hit a limit, increase or decrease size
+            if (change_buffer > REQ_LIMIT_INC) {
+                new_buffer_size = nextBuffer.length * 2;
+            } else if (change_buffer < REQ_LIMIT_DEC) {
+                new_buffer_size = nextBuffer.length / 2;
+            }
+
+            // if buffer size is to be changed, replace nextBuffer
+            if (new_buffer_size > 0) {
+                if ((MIN_BUFFER_SIZE <= new_buffer_size) &&
+                        (new_buffer_size <= MAX_BUFFER_SIZE)) {
+                    nextBuffer = new byte[new_buffer_size];
+                    if (debug != null) {
+                        debug.println("Buffer size changed to " +
+                                new_buffer_size);
+                    }
+                } else {
+                    if (debug != null) {
+                        debug.println("Buffer reached limit: " +
+                                nextBuffer.length);
+                    }
+                }
+                change_buffer = 0;
+            }
+
+            // load fresh random bytes into nextBuffer
+            lastRead = time;
+            readFully(nextIn, nextBuffer);
+            buffered = nextBuffer.length;
+        }
+
+        private void implNextBytes(byte[] data) {
+            try {
+                int data_len = data.length;
+                int ofs = 0;
+
+                while (data_len > 0) {
+                    int len;
+
+                    synchronized (LOCK_GET_BYTES) {
+                        int buf_pos;
+
+                        ensureBufferValid();
+                        buf_pos = nextBuffer.length - buffered;
```

Comment:
Please limit the scope of locals to where they're needed:
```
                    synchronized (LOCK_GET_BYTES) {
                        ensureBufferValid();
                        int buf_pos = nextBuffer.length - buffered;
                        int len;
```

Reply:
updated